File: cpp/linked-list/medium/92.反转链表-ii.cpp
/*
 * @lc app=leetcode.cn id=92 lang=cpp
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *node = nullptr;
    ListNode *reverseBetween(ListNode *head, int left, int right)
    {
        ListNode dummy;
        dummy.next = head;
        auto p     = head;
        auto pre   = &dummy;
        for (int i = 1; i < left; i++) {
            pre = p;
            p   = p->next;
        }
        auto h    = reverseNList(p, right - left + 1);
        pre->next = h;

        return dummy.next;
    }

    ListNode *reverseNList(ListNode *head, int n)
    {
        if (n == 1) {
            node = head->next;
            return head;
        }

        auto newHead     = reverseNList(head->next, n - 1);
        head->next->next = head;
        head->next       = node;
        return newHead;
    }
};
// @lc code=end

File: golang/linked-list/medium/92.反转链表 II.go
/*
 * @lc app=leetcode.cn id=92 lang=golang
 * @lcpr version=30202
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	var successor *ListNode
	var reverseN func(*ListNode, int) *ListNode
	reverseN = func(head *ListNode, n int) *ListNode {
		if n == 1 {
			successor = head.Next
			return head
		}

		newHead := reverseN(head.Next, n-1)
		head.Next.Next = head
		head.Next = successor
		return newHead
	}

	if left == 1 {
		return reverseN(head, right)
	}

	newHead := reverseBetween(head.Next, left-1, right-1)
	head.Next = newHead
	return head
}

// @lc code=end

/*
// @lcpr case=start
// [1,2,3,4,5]\n2\n4\n
// @lcpr case=end

// @lcpr case=start
// [5]\n1\n1\n
// @lcpr case=end

*/

